src/Services/Device.vala

```diff
@@ -0,0 +1,342 @@
+        state = (State) upower_device.state;
+        technology = (Technology) upower_device.technology;
+        update_time = upower_device.update_time;
+
+        is_charging = state == State.FULLY_CHARGED || state == State.CHARGING;
+        is_a_battery = device_type != Type.UNKNOWN && device_type != Type.LINE_POWER;
+
+        properties_updated ();
+    }
+
+    public bool is_present () {
+        bool present = false;
+        if (upower.on_battery || upower_device.is_present) {
+            present = true;
+        }
+
+        return present;
+    }
+
+    public string get_info () {
+        var percent = (int)Math.round (percentage);
+        if (percent <= 0) {
+            return _("Calculating…");
+        }
+
+        if (percent == 100 && is_charging) {
+            return _("Fully charged");
+        }
+
+        var info = _("%i%% charged").printf (percent);
+
+        return info;
+    }
+
+    public string get_icon_name_for_battery () {
+        if (!is_a_battery) {
+            return "preferences-system-power-symbolic";
+        }
+        if (percentage == 100 && is_charging) {
+            return "battery-full-charged";
+        }
+        var battery_icon = get_battery_icon ();
+        if (is_charging) {
+            return battery_icon + "-charging";
+        } else {
+            return battery_icon;
+        }
+    }
+
+    public string get_battery_icon () {
+        if (percentage <= 0) {
+            return "battery-good";
+        }
+
+        if (percentage < 10 && (time_to_empty == 0 || time_to_empty < 30 * 60)) {
+            return "battery-empty";
+        }
+
+        if (percentage < 30) {
+            return "battery-caution";
+        }
+
+        if (percentage < 60) {
+            return "battery-low";
+        }
+
+        if (percentage < 80) {
+            return "battery-good";
+        }
+
+        return "battery-full";
+    }
+
+    public string get_current_charge () {
+        return format_capacity (percentage, "%");
+    }
+
+    public string get_max_capacity () {
+        return format_capacity (capacity, "%");
+    }
+
+    public string get_design_energy () {
+        return format_capacity (energy_full_design, " Wh");
+    }
+
+    private static string format_capacity (double value, string unit) {
+        var value_int = (int)Math.round (value);
+        if (value_int == 0) {
+            return _("Unknown");
+        }
+
+        return value_int.to_string () + unit;
+    }
+
+    public string get_health () {
+        var capacity = (int)Math.round (capacity);
+
+        if (capacity == 0) {
+            return _("Unknown");
+        }
+
+        if (capacity < 60) {
+            return _("Critical");
+        }
+
+        if (capacity < 70) {
+            return _("Poor");
+        }
+
+        if (capacity < 80) {
+            return _("Fair");
+        }
+
+        if (capacity < 90) {
+            return _("Good");
+        }
+
+        return _("Excellent");
```

Comment:
Maybe we should add `//TRANSLATORS:` comments here? These strings may be ambiguous if there are no context.

Reply:
```suggestion
        if (capacity == 0) {
            return _("Unknown");
        }

        if (capacity < 60) {
            ///TRANSLATORS: Battery capacity below 60%, considered critically low battery health
            return _("Critical");
        }

        if (capacity < 70) {
            ///TRANSLATORS: Battery capacity from 60 to 69%, considered poor battery health
            return _("Poor");
        }

        if (capacity < 80) {
            ///TRANSLATORS: Battery capacity from 70 to 79%, considered fair battery health
            return _("Fair");
        }

        if (capacity < 90) {
            ///TRANSLATORS: Battery capacity from 80 to 90%, considered good battery health
            return _("Good");
        }

        ///TRANSLATORS: Battery capacity above 90%, considered excellent battery health
        return _("Excellent");
```